feat: add addDevToolbarPlugin and addIntegration utilities (#64)
florian-lefebvre/astro-integration-kit: docs/src/content/docs/utilities/add-devtoolbar-plugin.mdx

---
title: addDevToolbarPlugin

> description:  allows you to register a framework component as a Dev Toolbar Plugin.

Reviewer: ```suggestion
description:  Allows you to register a framework component as a Dev Toolbar Plugin.
```

---
import { Tabs, TabItem } from '@astrojs/starlight/components';

> `addDevToolbarPlugin` allows you to register a framework component as a Dev Toolbar Plugin!
> 
> So now you can use a React, Preact, Solid, Vue or Svelte component as a Dev Toolbar Plugin,
> instead of having to create your DOM manually yourself!

Reviewer: ```suggestion
`addDevToolbarPlugin` allows you to register a framework component as a Dev Toolbar Plugin! You can now use a React, Preact, Solid, Vue or Svelte component instead of manipulating the DOM imperatively!
```

                }) => {
                    const { resolve } = createResolver(import.meta.url);

                    addIntegration(Vue())

                    addDevToolbarPlugin({
                        framework: "vue",
                        name: "Test Vue Plugin",
                        id: "my-vue-plugin",
                        icon: `<svg version="1.1" viewBox="0 0 261.76 226.69" xmlns="http://www.w3.org/2000/svg"><g transform="matrix(1.3333 0 0 -1.3333 -76.311 313.34)"><g transform="translate(178.06 235.01)"><path d="m0 0-22.669-39.264-22.669 39.264h-75.491l98.16-170.02 98.16 170.02z" fill="#41b883"/></g><g transform="translate(178.06 235.01)"><path d="m0 0-22.669-39.264-22.669 39.264h-36.227l58.896-102.01 58.896 102.01z" fill="#34495e"/></g></g></svg>`,
                        src: resolve("./my-plugin.vue"),
                        style: `
                            h1 {
                                font-family: Inter;
                            }
                        `,
                        callback: (canvas, window) => {},
                    });
                }
            }
        }
    })
    ```
    ```vue title="my-integration/my-plugin.vue"

>     <script>

Reviewer: just don't put scripts if they're empty, but I think it would be great to have a real simple counter (i think you can take one from the official vue docs)

                        src: resolve("./my-plugin.vue"),
                        style: `
                            h1 {
                                font-family: Inter;
                            }
                        `,
                        callback: (canvas, window) => {},
                        updateConfig,
                        addDevToolbarApp,
                        injectScript,
                    })
                }
            }
        }
    }
    ```
    ```vue title="integration/my-plugin.vue"
    <script>

    </script>

    <template>
        <h1>Room with a Vue</h1>
    </template>
    ```
    </TabItem>
</Tabs>

## Adding Framework-specific Integrations

It's recommended to manually add the official Astro integration for your chosen framework
using the [addIntegration](/utilities/add-integration) utility instead of relying on your users
to install and add it. `addIntegration` won't re-add integrations, so you're safe to call it even
if your user _has_ added it already.

> 

Reviewer: Would be great great to have a link card to `addIntegration` as well here

    setup() {
        const { resolve } = createResolver(import.meta.url);

        return {
            "astro:config:setup": ({
                addDevToolbarPlugin,
            }) => {
                addDevToolbarPlugin({
                    framework: "svelte",
                    name: "Test Svelte Plugin",
                    id: "my-svelte-plugin",
                    icon: `<svg version="1.1" viewBox="0 0 261.76 226.69" xmlns="http://www.w3.org/2000/svg"><g transform="matrix(1.3333 0 0 -1.3333 -76.311 313.34)"><g transform="translate(178.06 235.01)"><path d="m0 0-22.669-39.264-22.669 39.264h-75.491l98.16-170.02 98.16 170.02z" fill="#41b883"/></g><g transform="translate(178.06 235.01)"><path d="m0 0-22.669-39.264-22.669 39.264h-36.227l58.896-102.01 58.896 102.01z" fill="#34495e"/></g></g></svg>`,
                    src: resolve("./my-plugin.svelte"),
                    style: readFileSync(resolve('./styles.css'), 'utf8'),
                    callback: (canvas, window) => {},
                })
            }
        }
    }
})
```

#### Bonus 🎉

If you use the `watch` flag on the Tailwind CLI and use the [watchIntegration](/utilities/watch-integration)

> plugin, you can then get automatic reloading of your plugin. Not HMR yet though!

Reviewer: really? i'd expect it to reload because `style.css` is watched right?

Reviewer 2: It's an Astro bug that styles are not getting HMR-ed, I believe Matthew was working on it.
